File: Data Structures & Algorithms/count-connected-components/submission-2.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        adj = {node: [] for node in range(n)}
        visited = set()
        count = 0

        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
        
        def dfs(node, par):
            if node in visited:
                return
            
            visited.add(node)
            for nei in adj[node]:
                if nei == par:
                    continue
                dfs(nei, node)
        
        for node in range(n):
            if node in visited:
                continue
            dfs(node, -1)
            count += 1 
        
        return count

File: Data Structures & Algorithms/redundant-connection/submission-2.java
class Solution {
    private int cycleStart = -1;
    private Set<Integer> cycle = new HashSet<>();
    private Set<Integer> visited = new HashSet<>();
    private Map<Integer, List<Integer>> adj = new HashMap<>();

    public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length;
        for (int i = 0; i <= n; i++)
        {
            adj.put(i, new ArrayList<Integer>());
        }

        for (int[] pair : edges)
        {
            adj.get(pair[0]).add(pair[1]);
            adj.get(pair[1]).add(pair[0]);
        }

        dfs(1, -1);

        for (int i = n - 1; i >= 0; i--)
        {
            int[] pair = edges[i];
            if (cycle.contains(pair[0]) && cycle.contains(pair[1]))
            {
                return new int[] {pair[0], pair[1]};
            }
        }
        
        return new int[0];
    }

    public boolean dfs(int node, int par)
    {
        if (visited.contains(node))
        {
            cycleStart = node;
            return true;
        }

        visited.add(node);
        for (int nei : adj.get(node))
        {
            if (nei == par)
            {
                continue;
            }
            if (dfs(nei, node))
            {
                if (cycleStart != -1)
                {
                    cycle.add(node);
                }
                if (cycleStart == node)
                {
                    cycleStart = -1;
                }
                return true;
            }
        }
        return false;
    }
}
